pycytominer/cyto_utils/cells.py

```diff
@@ -430,46 +435,210 @@ def aggregate_compartment(
+                    .count()
+                    .reset_index()
+                    .rename(
+                        columns={
+                            f"{self.fields_of_view_feature}": f"Metadata_Site_Count"
+                        }
+                    )
                 )
-            )
 
-            object_df = fields_count_df.merge(object_df, on=self.strata, how="right")
+                partial_object_df = fields_count_df.merge(
+                    partial_object_df,
+                    on=self.strata,
+                    how="right",
+                )
+
+            object_dfs.append(partial_object_df)
+
+        # Concatenate one or more aggregated dataframes row-wise into final output
+        object_df = pd.concat(object_dfs, axis=0).reset_index(drop=True)
 
         return object_df
 
+    def _compartment_df_generator(
+        self,
+        compartment,
+        n_strata=1,
+    ):
+        """A generator function that returns chunks of the entire compartment
+        table from disk.
+
+        We want to return dataframes with all compartment entries within unique
+        combinations of self.merge_cols when aggregated by self.strata
+
+        Parameters
+        ----------
+        compartment : str
+            Compartment to aggregate.
+        n_strata : int, default 1
+            Number of unique strata to pull from the database into working memory
+            at once.  A larger number does not lead to faster compute times
+
+        Yields
+        ------
+        image_df : pandas.core.frame.DataFrame
+            A generator whose __next__() call returns a chunk of the compartment
+            table, where rows comprising a unique aggregation stratum are not split
+            between chunks, and thus groupby aggregations are valid
+
+        """
+
+        assert (
+            n_strata > 0
+        ), "Number of strata to pull into memory at once (n_strata) must be > 0"
+
+        # Columns of the compartment table we will obtain
+        cols = "*"
+        compartment_row1 = pd.read_sql(
+            sql=f"select {cols} from {compartment} limit 1",
+            con=self.conn,
+        )
+        typeof_str = ", ".join([f"typeof({x})" for x in compartment_row1.columns])
+        compartment_dtypes = pd.read_sql(
+            sql=f"select {typeof_str} from {compartment} limit 1",
+            con=self.conn,
+        )
+        dtype_dict = dict(
+            zip(
+                [s[7:-1] for s in compartment_dtypes.columns],  # strip typeof( )
+                compartment_dtypes.iloc[0].values,
+            )
+        )
+
+        # Obtain all valid strata combinations, and their merge_cols values
+        df_unique_mergecols = (
+            self.image_df[self.strata + self.merge_cols]
+            .groupby(self.strata)
+            .agg(lambda s: np.unique(s).tolist())
+            .reset_index(drop=True)
+        )
+
+        def _sqlite_strata_conditions(df, dtypes, n=1):
```

Comment:
Based on my read, this function is the most critical contribution in this PR (cc @niranjchandrasekaran). It's probably because I'm not an expert in generator functions and how they're used, but I am having some trouble following the logic.

I make a couple comments in this function to try to help my understanding, but if you can see where my knowledge gap is, please let me know. Right now, I think this function will be difficult to maintain unless we provide a couple more tests.

Reply:
I agree that there should be some dedicated tests to demonstrate the function of this function